include/envoy/network/connection.h

```diff
@@ -313,6 +313,11 @@ class ClientConnection : public virtual Connection {
    * registered via addConnectionCallbacks().
    */
   virtual void connect() PURE;
+
+  /**
+   * Returns true if the underlying transport is secure.
+   */
+  virtual bool isSecureTransport() const PURE;
```

Comment:
is this only for health checker? If so I think it might make more sense to let `createHealthCheckConnection` return whether the connection is secured?

Reply:
Health checker uses the codec_client as member, and bind a connection instance with that.

`createHealthCheckConnection` is only invoked during [initialization](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/health_checker_impl.cc#L200, but later on every request we need to do this `setupUpstreamScheme` thing, with only access to [`codec_client_`](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/health_checker_impl.cc#L208)

Reply:
I know, but you can cache the value from connection creation there. That won't change after connection initialization.

Reply:
> cache the value from connection there,

do you mean cache into `health_check_impl` itself? or codec_client?

Trying to understand two comments together, is the end goal to not adding `implementsSecureTransport` to both `TransportSocketFactory`? Which is kinda redundant?

Reply:
in short term (while health checker is not using connpool, there's a TODO there) have a flag in `health_check_impl` should be enough.

Reply:
not needed since now we make `transportSocketFactory` part of `HostDesription`, we'll just invoke that directly.